bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/Desugar.scala

```diff
@@ -187,7 +187,39 @@ object Desugar extends Transformer with TreeDSL {
       treeCopy.MatchStatement(expr, value,
           clauses map(c => matchExpressionClauseToBindStatement(result, c)),
           exprToBindStatement(result, elseExpression))
-    
+
+    case Record(label, fields) =>
+      var unnested: Option[(Variable, Expression)] = None
+      def defer(expr: Expression) = {
+        val newVar = Variable.newSynthetic()
+        unnested = Some((newVar, expr))
+        newVar
+      }
+      
+      val newFields = fields.reverse.map { field =>
+        if (unnested.isDefined) {
+          field
+        } else {
+          treeCopy.RecordField(field, field.feature, field.value match {
+            case expr: StatAndExpression => defer(expr)
+            case expr: LocalExpression   => defer(expr)
+            case expr: CallExpression    => defer(expr)
+            case expr: MatchExpression   => defer(expr)
+            case expr: IfExpression      => defer(expr)
+            case expr: StatAndExpression => defer(expr)
+            case expr: Record            => defer(expr)
```

Comment:
How is this list made?
Many of those can contain arbitrary expressions,
so the idea is to allow everything but those after the tail call?

I think the opposite list would be safer: what is "pure" and be delayed without the user noticing: numbers, literals, records of literals, etc.